core/src/main/java/org/mule/util/metadata/ResultFactory.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.util.metadata;
+
+import org.mule.api.metadata.FailureType;
+import org.mule.api.metadata.Result;
+import org.mule.extension.api.metadata.MetadataResolvingException;
+import org.mule.util.ExceptionUtils;
+
+public final class ResultFactory
+{
+
+    public static <T> Result<T> success(T payload)
+    {
+        return new Result<T>()
+        {
+            @Override
+            public T get()
+            {
+                return payload;
+            }
+
+            @Override
+            public boolean isSucess()
+            {
+                return true;
+            }
+
+            @Override
+            public String getMessage()
+            {
+                return "Success";
+            }
+
+            @Override
+            public FailureType getFailureType()
+            {
+                return FailureType.NONE;
+            }
+
+            @Override
+            public String getStacktrace()
+            {
+                return "";
+            }
+        };
+    }
+
+    public static <T> Result<T> failure(T newPayload, Result originalResult)
+    {
+        return failure(newPayload, originalResult.getMessage(), originalResult.getFailureType(), originalResult.getStacktrace());
+    }
+
+    public static <T> Result<T> failure(T payload, String message, Exception e)
+    {
+        return failure(payload, message, getExceptionFailure(e), ExceptionUtils.getStackTrace(e));
+    }
+
+    public static <T> Result<T> failure(T payload, String message, FailureType failure, String stackTrace)
+    {
+        return new Result<T>()
+        {
+
+            @Override
+            public T get()
+            {
+                return payload;
+            }
+
+            @Override
+            public boolean isSucess()
+            {
+                return false;
+            }
+
+            @Override
+            public String getMessage()
+            {
+                return message;
+            }
+
+            @Override
+            public FailureType getFailureType()
+            {
+                return failure;
+            }
+
+            @Override
+            public String getStacktrace()
+            {
+                return stackTrace;
+            }
+        };
+    }
+
+    public static <T> Result<T> mergeResults(T payload, Result first, Result second)
+    {
+        if (first.isSucess() && second.isSucess())
+        {
+            return ResultFactory.success(payload);
```

Comment:
Message should not care if the result was successful, just use the payload. This enforces the idea of Optional<MetadataFailure> in the result instead of having message, code and reason as separate fields.